libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/for_loop.hpp

```diff
@@ -0,0 +1,1369 @@
+    typename util::detail::algorithm_result<ExPolicy>::type for_loop_n_strided(
+        ExPolicy&& policy, Rng&& rng, Size size, Args&&... args);
+}}    // namespace hpx::ranges
+
+#else
+
+#include <hpx/config.hpp>
+#include <hpx/assert.hpp>
+#include <hpx/functional/tag_invoke.hpp>
+#include <hpx/modules/concepts.hpp>
+#include <hpx/modules/executors.hpp>
+#include <hpx/modules/iterator_support.hpp>
+#include <hpx/modules/type_support.hpp>
+#include <hpx/parallel/algorithms/for_loop.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace ranges {
+    HPX_INLINE_CONSTEXPR_VARIABLE struct for_loop_t final
+      : hpx::functional::tag<for_loop_t>
+    {
+    private:
+        // clang-format off
+        template <typename ExPolicy, typename Iter, typename Sent, typename... Args,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_iterator<Iter>::value &&
+                hpx::traits::is_sentinel_for<Sent, Iter>::value
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy>::type
+        tag_invoke(hpx::ranges::for_loop_t, ExPolicy&& policy, Iter first,
+            Sent last, Args&&... args)
+        {
+            static_assert(sizeof...(Args) >= 1,
+                "for_loop must be called with at least a function object");
+
+            using hpx::util::make_index_pack;
+            return parallel::v2::detail::for_loop(
+                std::forward<ExPolicy>(policy), first, last, 1,
+                typename make_index_pack<sizeof...(Args) - 1>::type(),
+                std::forward<Args>(args)...);
+        }
+
+        // clang-format off
+        template <typename Iter, typename Sent, typename... Args,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::traits::is_iterator<Iter>::value &&
+                hpx::traits::is_sentinel_for<Sent, Iter>::value
+            )>
+        // clang-format on
+        friend void tag_invoke(
+            hpx::ranges::for_loop_t, Iter first, Sent last, Args&&... args)
+        {
+            static_assert(sizeof...(Args) >= 1,
+                "for_loop must be called with at least a function object");
+
+            return for_loop(parallel::execution::seq, first, last,
+                std::forward<Args>(args)...);
+        }
+
+        // clang-format off
+        template <typename ExPolicy, typename R, typename... Args,
+            HPX_CONCEPT_REQUIRES_(
+                hpx::is_execution_policy<ExPolicy>::value &&
+                hpx::traits::is_range<R>::value
+            )>
+        // clang-format on
+        friend typename parallel::util::detail::algorithm_result<ExPolicy>::type
+        tag_invoke(
+            hpx::ranges::for_loop_t, ExPolicy&& policy, R&& rng, Args&&... args)
```

Comment:
Is a pair of integers satisfying all requirement of a range?

Reply:
Should it? This would have the same behaviour as `std::vector<T> c(...); hpx::for_loop(std::begin(c), std::end(c), ...);`, i.e. it just passes the iterator to the element-wise function.

Reply:
Or do you actually mean if we should accept `std::pair<Integral>` as a range?

Reply:
The original `for_loop` supports being used with integral arguments: `for_loop(0, 10, [](int i) {});`. If the proposed `ranges::for_loop` does not support this (and as a pair of integers does not satisfy the requirements of a range, see http://eel.is/c++draft/range.range#3.3) I tend to think that we shouldn't add a corresponding overload.

Reply:
I'm confused. Is your concern that `hpx::for_loop` accepts integral indices *or* iterators, and as such `hpx::ranges::for_loop` should as well support something analogous, but not only ranges?

`hpx::for_loop` supports iterators (and integral arguments). `hpx::ranges::for_loop` supports iterator/sentinel pairs and ranges. This to me is analogous to `hpx::for_each` and `hpx::ranges::for_each` except for the additional support for integral arguments on `hpx::for_loop`.

Reply:
> Is your concern that `hpx::for_loop` accepts integral indices _or_ iterators, and as such `hpx::ranges::for_loop` should as well support something analogous, but not only ranges?

Yes, that is my concern. OTOH, I understand where you're coming from. 

Originally `hpx::for_loop` (based on [N4755](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4755.pdf)) was meant to provide an alternative to OMP parallel for loops in C++, i.e. was meant to work with integral types most and foremost. It just turned out that iterators fulfill all of the requirements such an algorithm imposes onto its integral (loop boundary) arguments, which lead to the unwieldy thing we have today that accepts both, integral and iterator types. As such, reducing the `hpx::ranges::for_loop` to supporting just iterators does not 'feel' right.

Reply:
Ok, I see your point of view as well. In that case wouldn't all of the ranges overloads for `for_loop` be slightly 'wrong', not just the range but also iterator/sentinel overloads...? Again, I'm mostly indifferent, but I think this makes `for_loop` a nice generalization of `for_each` so I'm slightly (but only slightly) in favor of keeping this (but then I opened the PR, so that shouldn't be a surprise).

There is no update to the parallelism ts after the ranges changes, right? I.e. currently there are no proposals for what, if any, overloads there should be for `hpx::ranges::for_loop`? Then it's a question of if we provide these overloads just because we can, or if we are conservative and stick to only what's proposed/standardized.

Reply:
> There is no update to the parallelism ts after the ranges changes, right? 

None that I know of.

> Then it's a question of if we provide these overloads just because we can, or if we are conservative and stick to only what's proposed/standardized.

All of our ranges overloads that take an execution policy are 'because we can'. It's way too late to play the conservative card now ;-)

Let's have your overloads then. I don't want to be in the way of progress.

Reply:
No problem :P I don't *really* badly need these but I think they are nice enough to include. Then we go with this. Thanks! And thanks for being appropriately skeptical, I hadn't thought about these issues so it's good to be on the same page.